tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +182,217 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
-    if (this != &other) {
-        this->device_ = other.device_;
-        this->size_ = other.size_;
-        this->address_ = other.address_;
-        this->page_size_ = other.page_size_;
-        this->buffer_type_ = other.buffer_type_;
-        this->buffer_layout_ = other.buffer_layout_;
-        this->shard_parameters_ = std::move(other.shard_parameters_);
-        this->bottom_up_ = other.bottom_up_;
-        this->buffer_page_mapping_ = std::move(other.buffer_page_mapping_);
-        this->allocate_ = other.allocate_;
-        // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-        // transferred to `this`
-        other.device_ = nullptr;
+}
+
+void Buffer::deallocate() {
+    if (!prepare_deallocation(allocation_status_)) {
+        return;
+    }
+
+    device_->push_work([self = weak_self.lock()] {
+        auto expected_status = AllocationStatus::DEALLOCATION_REQUESTED;
+        if (!self->allocation_status_.compare_exchange_strong(expected_status, AllocationStatus::DEALLOCATED)) {
+            // Buffer was already deallocated, nothing to do
+            return;
+        }
+
+        if (self->device_->initialized_ && self->size_ != 0) {
+            detail::BUFFER_MAP.erase({self->device_->id(), self->address_});
+            detail::DeallocateBuffer(self.get());
+        }
+    });
+}
+
+void Buffer::deleter(Buffer* buffer) {
+    // There is no concurrent allocations/deallocations happening, so no extra checks are required
+    if (buffer->allocation_status_ == AllocationStatus::DEALLOCATED) {
+        return;
+    }
+
+    buffer->device_->push_work([buffer] {
+        std::unique_ptr<Buffer> unique_buffer = std::unique_ptr<Buffer>(buffer);
+
+        if (!buffer->device_->initialized_ || buffer->size_ == 0) {
+            return;
+        }
+
+        detail::BUFFER_MAP.erase({buffer->device_->id(), buffer->address_});
+        detail::DeallocateBuffer(buffer);
+    });
+}
+
+bool Buffer::is_allocated() const {
+    auto allocation_status = allocation_status_.load();
+
+    if (device_->can_use_passthrough_scheduling()) {
+        return allocation_status == AllocationStatus::ALLOCATED;
+    }
+
+    // For calls from different threads we consider buffer to be allocated even if it's just ALLOCATION_REQUESTED or ALLOCATING,
+    // because once the caller will try to access it, the buffer will already be fully allocated
+    return allocation_status == AllocationStatus::ALLOCATION_REQUESTED
+        || allocation_status == AllocationStatus::ALLOCATING
+        || allocation_status == AllocationStatus::ALLOCATED;
+}
+
+uint32_t Buffer::address() const {
+    if (device_->can_use_passthrough_scheduling()) {
+        return address_;
+    }
+
+    // Waiting for the buffer to be allocated if the allocation is pending
+    allocation_status_.wait(AllocationStatus::ALLOCATION_REQUESTED);
+    allocation_status_.wait(AllocationStatus::ALLOCATING);
```

Comment:
what if we started waiting too late and missed ALLOCATION_REQUESTED?

Reply:
So we first wait for the status not to be ALLOCATION_REQUESTED, and then not to be ALLOCATING. Because the states go only in one direction, missing the statuses is perfectly fine.

Reply:
I mean if we come to the first line and AllocationStatus::ALLOCATING, is there a possibility to wait forever for ALLOCATION_REQUESTED?

Reply:
The syntax of `std::atomic::wait` is a bit reversed. This means to wait until `allocation_status_` is NOT ALLOCATION_REQUESTED.
https://en.cppreference.com/w/cpp/atomic/atomic/wait